dask_sql/physical/rex/core/call.py

```diff
@@ -689,41 +689,17 @@ def timestampadd(self, unit, interval, df: SeriesOrScalar):
         df = df.astype("datetime64[ns]")
 
         if unit in {"YEAR", "YEARS"}:
-            result = []
-            for date in df:
-                year = date.year + interval
-                # Check leap day
-                if year % 4 != 0 and date.month == 2 and date.day == 29:
-                    result.append(date.replace(year=year, month=3, day=1))
-                else:
-                    result.append(date.replace(year=year))
-            return pd.Series(result)
+            year = interval * 365
+            return df + timedelta(days=year)
         elif unit in {"QUARTER", "QUARTERS", "MONTH", "MONTHS"}:
-            result = []
-            for date in df:
-                if unit in {"QUARTER", "QUARTERS"}:
-                    month = date.month + (3 * interval)
-                else:  # "MONTH"
-                    month = date.month + interval
-                year = date.year
-                if month > 12:
-                    year = year + (month // 12)
-                    month = month % 12
-                # Check leap day
-                if year % 4 != 0 and month == 2 and date.day == 29:
-                    result.append(date.replace(year=year, month=3, day=1))
-                # Replace February 30 with March 2
-                elif month == 2 and date.day == 30:
-                    result.append(date.replace(year=year, month=3, day=2))
-                # Replace February 31 with March 3
-                elif month == 2 and date.day == 31:
-                    result.append(date.replace(year=year, month=3, day=3))
-                # Check months with 30 days
-                elif month in [4, 6, 9, 11] and date.day == 31:
-                    result.append(date.replace(year=year, month=month + 1, day=1))
-                else:
-                    result.append(date.replace(year=year, month=month))
-            return pd.Series(result)
+            if unit in {"QUARTER", "QUARTERS"}:
+                avg_days_in_quarter = 3 * ((30 * 4) + 28 + (31 * 7)) / 12
+                quarter = np.ceil(interval * avg_days_in_quarter)
```

Comment:
The only thing about this solution is that because we aren't iterating through the rows, we can't check that the days stay consistent, so there's a rounding error. I tried seeing if we could use some sort of `replace` function with the days, but was having trouble getting it to work.

For example:
```
SELECT
    TIMESTAMPADD(MONTH, 1, d) as "plus_1_month",
    TIMESTAMPADD(MONTH, 2, d) as "plus_2_months",
    TIMESTAMPADD(MONTH, 6, d) as "plus_6_months",
    TIMESTAMPADD(MONTH, 36, d) as "plus_36_months",
    TIMESTAMPADD(QUARTER, 1, d) as "plus_1_qt",
    TIMESTAMPADD(QUARTER, 2, d) as "plus_2_qts",
    TIMESTAMPADD(QUARTER, 6, d) as "plus_6_qts",
    TIMESTAMPADD(QUARTER, 36, d) as "plus_36_qts"
FROM df
```

produces:
<html>
<body>
<!--StartFragment-->

  | plus_1_month | plus_2_months | plus_6_months | plus_36_months | plus_1_qt | plus_2_qts | plus_6_qts | plus_36_qts
-- | -- | -- | -- | -- | -- | -- | -- | --
0 | 2021-11-03 15:53:42.000047 | 2021-12-03 15:53:42.000047 | 2022-04-04 15:53:42.000047 | 2024-10-02 15:53:42.000047 | 2022-01-03 15:53:42.000047 | 2022-04-04 15:53:42.000047 | 2023-04-04 15:53:42.000047 | 2030-10-01 15:53:42.000047

<!--EndFragment-->
</body>
</html>

While the *actual* expected output would be:
<html>
<body>
<!--StartFragment-->

  | plus_1_month | plus_2_months | plus_6_months | plus_36_months | plus_1_qt | plus_2_qts | plus_6_qts | plus_36_qts
-- | -- | -- | -- | -- | -- | -- | -- | --
0 | 2021-11-03 15:53:42.000047 | 2021-12-03 15:53:42.000047 | 2022-04-03 15:53:42.000047 | 2024-10-03 15:53:42.000047 | 2022-01-03 15:53:42.000047 | 2022-04-03 15:53:42.000047 | 2023-04-03 15:53:42.000047 | 2030-10-03 15:53:42.000047

<!--EndFragment-->
</body>
</html>

So the question is how much do we care about rounding errors when it comes to days? I'd also be happy to chat about other possible solutions.

Reply:
Haven't tried this out locally, but have you seen [`DateOffsets`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.tseries.offsets.DateOffset.html#pandas.tseries.offsets.DateOffset)? These seem to function identically to timedeltas but can be created using weeks/months/years as units, which should hopefully resolve the rounding error.

Reply:
That worked for me, thanks!